packages/stentor-conditional/src/ConditionalDeterminer.ts

```diff
@@ -0,0 +1,141 @@
+ * checks it can perform and then determines which is the best match.
+ */
+export class ConditionalDeterminer {
+
+    private timeout = 500;
+
+    private checks: ConditionalCheck[] = [];
+
+    public constructor(checks?: ConditionalCheck[]) {
+        if (Array.isArray(checks)) {
+            this.checks = checks;
+        }
+    }
+
+    /**
+     * Determine the best conditional matches
+     * 
+     * @param conditionals 
+     * @returns All matching conditionals based on the available checks
+     */
+    public determine<T>(conditionals: Conditional<T>[]): Conditional<T>[] {
+
+        if (!Array.isArray(conditionals)) {
+            return [];
+        }
+
+        if (conditionals.length === 0) {
+            return [];
+        }
+
+        const possible: Conditional<T>[] = [];
+
+        // For each of the conditionals
+        conditionals.forEach((conditional) => {
+            if (typeof conditional.conditions === "string") {
+                // Get all the function and add them to the sandbox.
+                const sandboxFunctions: { [name: string]: ((input: any) => boolean) } = {};
+
+                if (Array.isArray(this.checks)) {
+                    this.checks.forEach((check) => {
+                        if (Array.isArray(check.functions)) {
+                            check.functions.forEach((func) => {
+                                sandboxFunctions[func.name] = func;
+                            })
+                        }
+                    });
+                }
+                const vm = new VM({
+                    timeout: this.timeout,
+                    sandbox: {
+                        ...sandboxFunctions
+                    }
+                });
+                let result: boolean;
+                try {
+                    result = vm.run(conditional.conditions);
+                } catch (e) {
+                    log().error(`Error evaluating conditions "${conditional.conditions}": ${e}`);
+                    result = false;
+                }
+                // It should be a boolean but need to double check
+                if (typeof result === "boolean" && result) {
+                    possible.push(conditional)
+                }
+            } else if (typeof conditional.conditions === "object") {
+                const { must, should } = conditional.conditions;
+                let mustsPassed = false;
+                let mustsFailed = false;
+
+                if (Array.isArray(must) && must.length > 0) {
+                    must.forEach((obj) => {
+                        if (this.pass(obj)) {
+                            mustsPassed = true;
+                        } else {
+                            mustsFailed = true;
+                        }
+                    });
+                } else {
+                    mustsPassed = true;
+                }
+
+                if (mustsPassed && !mustsFailed) {
```

Comment:
If mustsPassed is true then mustsFailed must be false and vice-versa. The two booleans is confusing.